### PR TITLE
unicode imports

### PR DESCRIPTION
**Pull request by da-woods**
This hopefully "finishes" the support for unicode identifiers. It builds off https://github.com/cython/cython/pull/3096, so I don't think there's much point in looking at it before that's finalized.

It adds ~~2 features~~ 1 feature:
1. ~~Support for unicode identifiers in C/C++ features such as structs and cppclasses. For structs used purely in Python I've mangled the names with punycode. For features that are exported/imported to C with "public" or "extern", I've translated the names to be \uXXXX escaped (without any mangling or normalization). Pretty much every modern C/C++ compiler supports unicode in identifiers in this form (only Clang supports it in raw form I think), so this this seems like the most compatible thing to do. I've trusted that the user knows what names they want and not performed any normalization for these (I don't think normalization is yet defined in C/C++ standards, so it's hard to do anything else).~~

2. Support for import of modules with unicode in their names. As in PEP 489 this is only supported from Python 3.5 onwards (but the Cython translation step should work fine in earlier versions...). The Python 2 filename handling in Cython.Build seems a bit fragile with unicode filenames (since most os.path functions require bytes) - I haven't made any serious attempt to fix that beyond getting my test-cases to work.

### CODE REVIEW

File: Cython/Compiler/ModuleNode.py [hunk 1]
```diff
@@ -773,7 +782,7 @@ def generate_filename_table(self, code):
                 if isabs(file_path):
                     file_path = basename(file_path)  # never include absolute paths
                 escaped_filename = file_path.replace("\\", "\\\\").replace('"', r'\"')
-                code.putln('"%s",' % escaped_filename)
+                code.putln('%s,' % EncodedString(escaped_filename).as_c_string_literal())
```
Comment (scoder) [inline Cython/Compiler/ModuleNode.py]: This will not work for non-UTF8 file names in Py2. Not sure right now if this is a real issue, but it probably is.
Why was this change needed?

File: Cython/Compiler/ModuleNode.py [hunk 2]
```diff
@@ -2318,12 +2327,22 @@ def generate_module_init_func(self, imported_modules, env, code):
         code.enter_cfunc_scope(self.scope)
         code.putln("")
         code.putln(UtilityCode.load_as_string("PyModInitFuncType", "ModuleSetupCode.c")[0])
-        header2 = "__Pyx_PyMODINIT_FUNC init%s(void)" % env.module_name
+        try:
+            env.module_name.encode('ascii')
+        except UnicodeEncodeError:
+            py2_mod_name = "PYTHON2_DOES_NOT_SUPPORT_UNICODE_MODULE_NAMES"
```
Comment (scoder) [inline Cython/Compiler/ModuleNode.py]: I think the `#error` below is enough. This name wouldn't normally be user visible, I think. Just use the same name part as in Py3 here.

File: Cython/Compiler/ModuleNode.py [hunk 3]
```diff
@@ -2665,12 +2685,13 @@ def generate_module_import_setup(self, env, code):
         fq_module_name = self.full_module_name
         if fq_module_name.endswith('.__init__'):
             fq_module_name = fq_module_name[:-len('.__init__')]
+        fq_module_name = EncodedString(fq_module_name).as_c_string_literal()
```
Comment (scoder) [inline Cython/Compiler/ModuleNode.py]: This is a different thing than the `fq_module_name` before and deserves a different variable name to reduce the risk of accidental misuse in the future.

File: Cython/Compiler/ModuleNode.py [hunk 4]
```diff
@@ -3120,6 +3157,11 @@ def generate_type_import_call(self, type, code, import_generator, error_code=Non
     def generate_type_ready_code(self, entry, code):
         Nodes.CClassDefNode.generate_type_ready_code(entry, code)
 
+    def module_is_main_name(self):
```
Comment (scoder) [inline Cython/Compiler/ModuleNode.py]: I was failing to understand the meaning of this function by its name several times further up, before I found it here and investigated what it does. Seems difficult to name it properly, but a better name might be `is_main_module_flag_cname()`.

File: Cython/Compiler/ModuleNode.py [hunk 5]
```diff
@@ -2759,8 +2780,18 @@ def generate_main_method(self, env, code):
                 main_method=Options.embed,
                 wmain_method=wmain))
 
+    def punycode_module_name(self, name):
+        # from PEP483
+        try:
+            name = b'_' + name.encode('ascii')
```
Comment (scoder) [inline Cython/Compiler/ModuleNode.py]: I don't think this underscore belongs into a method that claims to return the `punycode_module_name`.
Maybe the method should accept an additional `prefix` and then `build…` something?

File: Cython/Compiler/Parsing.py [hunk 6]
```diff
@@ -2920,6 +2920,8 @@ def p_c_simple_declarator(s, ctx, empty, is_type, cmethod_flag,
                             fatal=False)
                 name = name + ' ' + op
                 s.next()
+        name = EncodedString(name)
+        if cname is not None: cname = EncodedString(cname)
```
Comment (scoder) [inline Cython/Compiler/Parsing.py]: Could this be moved into `p_opt_cname`?

File: Cython/Compiler/Symtab.py [hunk 7]
```diff
@@ -428,6 +425,26 @@ def mangle_class_private_name(self, name):
         if self.parent_scope:
             return self.parent_scope.mangle_class_private_name(name)
         return name
+    
+    def nomangle(self, name):
+        r"""
+        Allows unicode names to be translated to valid C/C++ code
```
Comment (scoder) [inline Cython/Compiler/Symtab.py]: Honestly, I would leave this out for now, or at least move it to a separate PR. It seems to deserve a separate discussion.

File: Cython/StringIOTree.py [hunk 8]
```diff
@@ -66,7 +66,13 @@ def copyto(self, target):
             child.copyto(target)
         stream_content = self.stream.getvalue()
         if stream_content:
-            target.write(stream_content)
+            try:
+                target.write(stream_content)
+            except UnicodeEncodeError as e:
```
Comment (scoder) [inline Cython/StringIOTree.py]: What is this used for?

File: Cython/Utility/CConvert.pyx [hunk 9]
```diff
@@ -18,7 +18,7 @@ cdef {{struct_type}} {{funcname}}(obj) except *:
         value = obj['{{member.name}}']
     except KeyError:
         raise ValueError("No value specified for struct attribute '{{member.name}}'")
-    result.{{member.cname}} = value
+    result.{{member.name}} = value
```
Comment (scoder) [inline Cython/Utility/CConvert.pyx]: What?

File: tests/run/unicode_identifiers.pyx [hunk 10]
```diff
@@ -184,6 +195,32 @@ class NormalClassΓΓ(Γναμε2):
             pass
         return nestεd
 
+def unicode_kwarg(*,αrg):
```
Comment (scoder) [inline tests/run/unicode_identifiers.pyx]: Space.
Also, I think an example where the difference is easier to spot (in different fonts) than for `arg` versus `αrg` would be nice.

File: tests/run/unicode_imports.srctree [hunk 11]
```diff
@@ -0,0 +1,98 @@
...
+
+PYTHON setup.py build_ext --inplace
+PYTHON -m mydoctest
+
+########### mydoctest.py #######
+
+import sys
+if (sys.version_info[0] < 3 or
+    (sys.version_info[0] == 3 and sys.version_info[1] < 5)):
+    exit() # I'm struggling to restrict srctree tests to a single Python version
```
Comment (scoder) [inline tests/run/unicode_imports.srctree]: There is a blacklist named `VER_DEP_MODULES` in `runtests.py`.

File: tests/run/unicode_imports.srctree [hunk 12]
```diff
@@ -0,0 +1,98 @@
...
+from Cython.Build import cythonize
+
+files = ["mymoδ.pyx", "from_cy.pyx"]
+
+
+# For Python 2 and Python <= 3.4 just run pyx->c; 
+# don't compile the C file
+modules = cythonize(files)
+
+if sys.version_info[0] >= 3 and sys.version_info[1] >= 5:
```
Comment (scoder) [inline tests/run/unicode_imports.srctree]: What about Python 4.2?

File: Cython/Utility/CConvert.pyx (hunk 9 above)
Comment (da-woods) [reply Cython/Utility/CConvert.pyx]: This is generating Cython code so I believe it's correct to use the name, rather than the cname (although the vast majority of the time they're the same). Here's a non-unicode example of something that works now but didn't before:

```
cdef struct S:
    int x "not_x"
def f(d):
    """
    >>> f({ 'x': 1 })
    1
    """
    cdef S s = d
    print(s.x)
```

Although really this is related to the struct support (which is getting moved out of this PR) I'm going to leave this bit in because it fixes existing bugs.

File: Cython/Compiler/Main.py [hunk 13]
```diff
@@ -433,6 +446,16 @@ def create_default_resultobj(compilation_source, options):
 def run_pipeline(source, options, full_module_name=None, context=None):
     from . import Pipeline
 
+    # ensure that the inputs are unicode (for Python 2)
+    try:
+        source = source.decode("utf-8")
```
[the remark this answers is not in the document]
Comment (da-woods) [reply Cython/Compiler/Main.py]: `source` here is the source filename, not the source code.

File: Cython/Compiler/Main.py [hunk 14]
```diff
@@ -433,6 +446,16 @@ def create_default_resultobj(compilation_source, options):
...
+    except AttributeError:
+        pass
+    try:
+        full_module_name = full_module_name.decode("utf-8")
```
[the remark this answers is not in the document]
Comment (da-woods) [reply Cython/Compiler/Main.py]: I did have a look at changing filenames and module to unicode names earlier. Seemed to involve rewriting nearly all the code that handles resolving cimport paths, so in the end I decided to do it here. I agree that it ideally should be done earlier, but it actually doing it was turning out to be too nasty. I'll make the version-specific bit clearer though

File: Cython/Compiler/ModuleNode.py (hunk 1 above)
Comment (da-woods) [reply Cython/Compiler/ModuleNode.py]: It needs `as_c_string_literal` to avoid putting unicode characters in the C source. `replace` returns a plain `unicode` object rather than an `EncodedString` even if the input is an `EncodedString` so it does need converting.

It does pass the testsuite in Python2 (the majority of which is plain ASCII filenames) so I think it isn't a real issue.

File: Cython/Compiler/ModuleNode.py (hunk 2 above)
Comment (da-woods) [reply Cython/Compiler/ModuleNode.py]: OK. I've used a version with unicode characters dropped as the placeholder here instead

File: Cython/StringIOTree.py (hunk 8 above)
Comment (da-woods) [reply Cython/StringIOTree.py]: This was debugging code that was accidentally left in. Removed now

File: tests/run/unicode_imports.srctree (hunk 11 above)
Comment (da-woods) [reply tests/run/unicode_imports.srctree]: Good to know. In this case I don't actually want the blacklist since I do want to test the cythonize stage with lower Python versions (just not the C compilation or running). I've clarified the comment though

File: Cython/Compiler/ModuleNode.py [hunk 15]
```diff
@@ -229,7 +234,10 @@ def generate_public_declaration(self, entry, h_code, i_code):
                 entry.type.declaration_code(entry.cname, pyrex=1)))
 
     def api_name(self, env):
-        return env.qualified_name.replace(".", "__")
+        api_name = self.punycode_module_name(env.qualified_name)
+        if api_name.startswith('_'):
+            api_name = api_name[1:]
```
[the remark this answers is not in the document]
Comment (da-woods) [reply Cython/Compiler/ModuleNode.py]: It's now changed to be slightly different. I was trying not to change the api names generated for existing non-unicode modules. The version I've changed it to generated `import_name` for non-unicode names (as before) and `importU_punycodename` for unicode modules, which I think is reasonable. It's essentially because punycode_module_name returns something starting with "_"

File: Cython/Compiler/Parsing.py (hunk 6 above)
Comment (da-woods) [reply Cython/Compiler/Parsing.py]: It looks to be already be in cname. This is in the part I've now taken out of the PR so I'll investigate when preparing the other one.

File: Cython/Build/Dependencies.py [hunk 16]
```diff
@@ -783,6 +783,8 @@ class Extension_setuptools(Extension): pass
     create_extension = ctx.options.create_extension or default_create_extension
 
     for pattern in patterns:
+        if sys.version_info[0]==2 and isinstance(pattern, unicode):
+            pattern = encode_filename_in_py2(pattern)
```
[the remark this answers is not in the document]
Comment (da-woods) [reply Cython/Build/Dependencies.py]: The problem was that pattern can be an `Extension` object which `encode_filename_in_py2` can't cope with. I've changed it to not call it when passed an `Extension` which is probably clearer

**Comment by da-woods** [comment]
I've moved the C stuff out of this PR - I'll make a new PR with it in at some point.

I think I've addressed most of the issues raised

File: Cython/Compiler/ModuleNode.py [hunk 17]
```diff
@@ -2665,13 +2699,14 @@ def generate_module_import_setup(self, env, code):
         fq_module_name = self.full_module_name
         if fq_module_name.endswith('.__init__'):
             fq_module_name = fq_module_name[:-len('.__init__')]
+        fq_module_name_cstring = EncodedString(fq_module_name).as_c_string_literal()
```
Comment (scoder) [inline Cython/Compiler/ModuleNode.py]: As mentioned above, `self.full_module_name` should be an `EncodedString` already (for consistency reasons), and then it only needs to be rewrapped above if it actually gets modified here.

File: Cython/StringIOTree.py [hunk 18]
```diff
@@ -39,6 +39,7 @@
     from cStringIO import StringIO
 except ImportError:
     from io import StringIO
+import sys
```
Comment (scoder) [inline Cython/StringIOTree.py]: Left-over?

File: Cython/Utility/CConvert.pyx (hunk 9 above)
Comment (scoder) [reply Cython/Utility/CConvert.pyx]: Ah, right. Yes, that's correct. Did you added this example as a test below?

File: tests/run/unicode_imports.srctree [hunk 19]
```diff
@@ -0,0 +1,105 @@
...
+
+import sys
+from Cython.Build import cythonize
+
+files = ["mymoδ.pyx", "from_cy.pyx"]
+
+
+# For Python 2 and Python <= 3.4 just run pyx->c; 
+# don't compile the C file
+modules = cythonize(files)
```
Comment (scoder) [inline tests/run/unicode_imports.srctree]: :+1:

File: Cython/Compiler/Main.py (hunk 13 above)
[the remark this answers is not in the document]
Comment (da-woods) [reply Cython/Compiler/Main.py]: Done - should "module_name" be the same? I assume it shouldn't, but it is derived from the filename usually

File: Cython/Compiler/ModuleNode.py [hunk 20]
```diff
@@ -28,11 +29,18 @@
...
+from ..Utils import replace_suffix as utils_replace_suffix
 from .Code import UtilityCode, IncludeCode
-from .StringEncoding import EncodedString
+from .StringEncoding import EncodedString, encoded_string_or_bytes_literal
 from .Pythran import has_np_pythran
 
+
+def replace_suffix(path, newsuf):
+    x = utils_replace_suffix(path, newsuf)
+    return encoded_string_or_bytes_literal(x, sys.getfilesystemencoding())
```
[the remark this answers is not in the document]
Comment (da-woods) [reply Cython/Compiler/ModuleNode.py]: Done. The problem is that Utils.replace_suffix doesn't seem to know about EncodedString/BytesLiteral (I could import them but it seems like a design decision that it doesn't use anything from "Compiler"), so it's quite hard to deal with the file name encoding earlier

File: Cython/Compiler/ModuleNode.py [hunk 21]
```diff
@@ -208,7 +216,14 @@ def h_entries(entries, api=0, pxd=0):
             h_code.putln("/* It now returns a PyModuleDef instance instead of a PyModule instance. */")
             h_code.putln("")
             h_code.putln("#if PY_MAJOR_VERSION < 3")
-            h_code.putln("PyMODINIT_FUNC init%s(void);" % env.module_name)
+            try:
+                env.module_name.encode("ascii")
```
[the remark this answers is not in the document]
Comment (da-woods) [reply Cython/Compiler/ModuleNode.py]: I've added an `isascii` method to `EncodedString` and `BytesLiteral` instead (for Py2 only). I haven't rewritten the punycoding code since this is taken from PEP489 and it seems best to match that.

File: Cython/Utility/CConvert.pyx (hunk 9 above)
Comment (da-woods) [reply Cython/Utility/CConvert.pyx]: Have done now. Also added a test for conversion the other way (which wasn't broken, but if you're testing one way it seems worth testing the other)

**Comment by da-woods** [comment]
Updated again to address your comments.

File: Cython/Compiler/StringEncoding.py [hunk 22]
```diff
@@ -146,6 +146,16 @@ def as_c_string_literal(self):
             s = bytes_literal(self.byteencode(), self.encoding)
         return s.as_c_string_literal()
 
+    if sys.version_info[0] == 2:
+        def isascii(self):
```
Comment (scoder) [inline Cython/Compiler/StringEncoding.py]: Nice Idea. I didn't even think of making it a method, but sure, makes total sense.

**Comment by scoder** [comment]
I'm still getting failures on Windows, even after resolving several encoding issues and what not:
https://ci.appveyor.com/project/cython/cython/builds/29947270
Here's the branch that I've been working with:
https://github.com/cython/cython/pull/3194

@da-woods, could you try to find out what's going wrong there?

**Comment by da-woods** [comment]
I'll have a look...

**Comment by da-woods** [comment]
@scoder I think it's a distutils bug (or maybe setuptools) rather than a Cython bug. When I run it on Windows I get:
```
C:\Program Files (x86)\Microsoft Visual Studio\2019\BuildTools\VC\Tools\MSVC\14.24.28314\bin\HostX86\x64\link.exe /nologo /INCREMENTAL:NO /LTCG /DLL /MANIFEST:EMBED,ID=2 /MANIFESTUAC:NO /LIBPATH:C:\Users\XXXXX\Miniconda3\envs\py37\libs /LIBPATH:C:\Users\XXXXX\Miniconda3\envs\py37\PCbuild\amd64 "/LIBPATH:C:\Program Files (x86)\Microsoft Visual Studio\2019\BuildTools\VC\Tools\MSVC\14.24.28314\lib\x64" "/LIBPATH:C:\Program Files (x86)\Windows Kits\10\lib\10.0.18362.0\ucrt\x64" "/LIBPATH:C:\Program Files (x86)\Windows Kits\10\lib\10.0.18362.0\um\x64" /EXPORT:PyInit_mymoð build\temp.win-amd64-3.7\Release\mymoð.obj /OUT:C:\Users\XXXXX\Documents\cython-fix_srctree_tests_on_windows\TEST_TMP\run\unicode_imports\mymoð.cp37-win_amd64.pyd /IMPLIB:build\temp.win-amd64-3.7\Release\mymoð.cp37-win_amd64.lib
LINK : error LNK2001: unresolved external symbol PyInit_mymoð
build\temp.win-amd64-3.7\Release\mymoð.cp37-win_amd64.lib : fatal error LNK1120: 1 unresolved externals
```

The key bit is `/EXPORT:PyInit_mymoð` - distutils/setuptools adds this to ensure that the `PyInit_` symbol has the correct linkage (presumably incase `PyMODINIT_FUNC ` has been forgotten). However, for a unicode filename the symbol is `PyInitU_<punicodified_something>`, so you get a link error.

Relevant code is in... 
https://github.com/python/cpython/blob/0d30ae1a03102de07758650af9243fd31211325a/Lib/distutils/command/build_ext.py#L686

I'll try to submit a bug report to the CPython project so they can fix it at source. However, in the short-term this looks like something we could patch in `cythonize` so I'll have a go at doing that.

**Comment by da-woods** [comment]
@scoder A really hacky patch for the Cython side of it:

```
    class DummyExportSymbols(list):
        # export_symbols for unicode filename cause link errors on Windows
        # Cython doesn't need them (it already defines PyInit with the correct linkage)
        # so use this class as a temporary fix to stop them from being generated
        def __init__(self):
            super(DummyExportSymbols, self).__init__() # empty list
        def __contains__(self, val):
            # so distutils doesn't "helpfully" add PyInit_<name>
            return True
    for m in module_list:
        m.export_symbols = DummyExportSymbols()
```

This goes immediately before this line: https://github.com/cython/cython/blob/074362b47093febfe5273ea7a00f4ba5ded6e73f/Cython/Build/Dependencies.py#L975. Posting it here because I don't know how to submit a PR against your branch.
